File: leetcode/4Sum/python/solution.py
class Solution:
    def fourSum(self, nums: list[int], target: int) -> list[list[int]]:
        """
        Time complexity: O(n3)
        Auxiliary space complexity: O(n)
        Tags:
            A: two pointers
        """
        nums.sort()
        N = len(nums)
        res = []

        for i in range(N - 3):
            # if not the first i loop
            # skip repeating nums
            if i > 0 and nums[i - 1] == nums[i]:
                continue
            
            for j in range(i + 1, N - 2):
                # if not the first j loop
                # skip repeating nums
                if j > i + 1 and nums[j - 1] == nums[j]:
                    continue
                
                left = j + 1
                right = N - 1
                
                while left < right:
                    quad = nums[i] + nums[j] + nums[left] + nums[right]

                    if target == quad:
                        res.append([nums[i], nums[j], nums[left], nums[right]])
                        left += 1
                        right -= 1
                        
                        while left < right and nums[left - 1] == nums[left]:
                            left += 1
                        
                    elif target < quad:
                        right -= 1
                    else:
                        left += 1

        return res


print(sorted(Solution().fourSum([1, 0, -1, 0, -2, 2], 0)) == sorted([[-2, -1, 1, 2], [-2, 0, 0, 2], [-1, 0, 0, 1]]))
print(sorted(Solution().fourSum([2, 2, 2, 2, 2], 8)) == sorted([[2, 2, 2, 2]]))
print(sorted(Solution().fourSum([0, 0, 0, 0], 0)) == sorted([[0, 0, 0, 0]]))
print(sorted(Solution().fourSum([1, -2, -5, -4, -3, 3, 3, 5], -11)) == sorted([[-5, -4, -3, 1]]))

File: leetcode/4Sum/javascript/solution.js
class Solution {
   /**
    * Time complexity: O(n3)
    * Auxiliary space complexity: O(n)
    * Tags:
    *    A: two pointers
    * @param {number[]} nums
    * @param {number} target
    * @return {}
    */
   fourSum(nums, target) {
      nums.sort((a, b) => a - b);
      const quadruplets = [];

      for (let i = 0; i < nums.length - 3; i++) {
         if (nums[i - 1] === nums[i])
            continue

         for (let j = i + 1; j < nums.length - 2; j++) {
            if (j > i + 1 && nums[j - 1] === nums[j])
               continue

            let left = j + 1;
            let right = nums.length - 1;

            while (left < right) {
               const quadruplet = nums[i] + nums[j] + nums[left] + nums[right]
               
               if (quadruplet === target) {
                  quadruplets.push([nums[i], nums[j], nums[left], nums[right]])
                  left++;
                  right--;
                  while (left < right && nums[left - 1] === nums[left])
                     left++;
               } else if (quadruplet < target)
                  left++;
               else
                  right--;
            }
         }
      }
      return quadruplets
   };
}


const fourSum = new Solution().fourSum;
console.log(JSON.stringify(new Solution().fourSum([1, 0, -1, 0, -2, 2], 0)) === JSON.stringify([[-2, -1, 1, 2], [-2, 0, 0, 2], [-1, 0, 0, 1]]))
console.log(JSON.stringify(new Solution().fourSum([2, 2, 2, 2, 2], 8)) === JSON.stringify([[2, 2, 2, 2]]))
console.log(JSON.stringify(new Solution().fourSum([0, 0, 0, 0], 0)) === JSON.stringify([[0, 0, 0, 0]]))
console.log(JSON.stringify(new Solution().fourSum([1, -2, -5, -4, -3, 3, 3, 5], -11)) === JSON.stringify([[-5, -4, -3, 1]]))
